### ISSUE TITLE
enlarge overlay info?

### ISSUE THREAD

**Issue by stsp** [tkchia/gcc-ia16#134]
I tried to add an overlay info into
an offset 0x1c of an executable,
but it seems to be already taken.
If the value is changed, program
hangs on start.
What is it currently used for?
And if its busy, would it be possible
to enlarge the header size by 1-2
paragraphs to leave some space
for user's overlay info?

**Comment by stsp** [comment]
This is the -mdosx32 mode.

**Comment by tkchia** [comment]
Hello @stsp,

> I tried to add an overlay info into an offset 0x1c of an executable,

That is where the `MZ` relocation table (currently) hangs out.

> And if its busy, would it be possible to enlarge the header size by 1-2 paragraphs to leave some space for user's overlay info?

Maybe.  I have some ideas about how to do this properly &mdash; without confusing the CauseWay stub loader, and (preferably) without messing up the overall file structure design.

Thank you!

**Comment by stsp** [comment]
> That is where the MZ relocation table (currently) hangs out.

OK I see, there seem to be just 1
relocation entry, and it starts exactly
where overlay info starts...

Anyway, looking forward to get it in any
proper way, thanks.

**Comment by tkchia** [comment]
Hello @stsp,

I just wrote a program [`lfanew`](https://gitlab.com/tkchia/lfanew) that can rewrite an `MZ` executable file to include a [`.e_lfanew`](https://learn.microsoft.com/en-us/archive/msdn-magazine/2002/february/inside-windows-win32-portable-executable-file-format-in-detail) header field pointing to a "new executable" payload.

(**Note:** this field is at **offset `0x3c`**, _not_ at offset `0x1c`.  It is a 32-bit field.)

So in effect this allows you to create an `MZ` file with _two_ overlays after the `MZ` part:
* an older-style overlay pointed to by `.e_cp` and `.e_cblp` (page count, last page bytes left) header fields;
* and a "new executable" overlay pointed to by the `.e_lfanew` field.

I also modified the CauseWay extender to fall back on reading the older-style overlay if it fails to parse the "new executable" overlay (it recognizes both types).

So now you can do something like this:
```
$ sudo apt-get install lfanew
...
$ ia16-elf-gcc -mdosx32 -Os old.c -o old.exe
.../ia16-elf/bin/ld.gold: warning: script places BSS section in the middle of a LOAD segment; space will be
 allocated in the file
$ lfanew -o old.fix old.exe
$ hexdump -C -v old.fix | head -10
00000000  4d 5a 2c 01 4f 00 00 00  04 00 24 0e ff ff d3 17  |MZ,.O.....$.....|
00000010  00 02 00 00 00 00 00 00  40 00 00 00 00 00 00 00  |........@.......|
00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 20 a1 00 00  |............ ...|
00000040  b9 4b 4e be 94 9c 89 f7  1e a9 b5 80 8c c8 05 05  |.KN.............|
00000050  00 8e d8 05 14 0e 8e c0  fd f3 a5 fc 2e 80 6c 12  |..............l.|
00000060  10 73 e7 92 af ad 0e 0e  0e 06 1f 07 16 bd 0f 00  |.s..............|
00000070  bb ef 80 55 cb 55 50 58  21 0d 03 09 09 ac dc 1b  |...U.UPX!.......|
00000080  f6 f3 95 9c 9e 1b 7d 01  ae 9b 00 3e 7d 01 00 43  |......}....>}..C|
00000090  ff 43 61 75 73 65 57 61  79 ff 20 44 4f 53 20 45  |.CauseWay. DOS E|
$ hexdump -C -v old.fix | tail -5
0000a0e0  64 21 00 43 00 00 00 00  f8 00 40 00 00 00 00 00  |d!.C......@.....|
0000a0f0  00 00 00 00 00 00 00 00  00 00 00 00 37 00 00 00  |............7...|
0000a100  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0000a110  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0000a120
$ cat old.fix new.coff >new.exe
```
(In this case the old overlay will start at offset `(0x4f - 1) * 0x200 + 0x12c` = `0x9d2c`, and the new overlay at offset `0xa120`.)

Thank you!

**Comment by stsp** [comment]
Thanks!
So you added the payload offset
to 0x3c and also vacated 0x1c for
me - this is exactly what  I needed.
I pushed the support for that scheme -
it was a major stopper for me before.

Though the requirement of a new tool
looks unexpected. Maybe you'll consider
making it a default (or optional via a
switch) format of ia16-gcc. But I can
live with a new tool either, if this is an
only option.

**Comment by tkchia** [comment]
Hello @stsp,

> So you added the payload offset to 0x3c and also vacated 0x1c for me - this is exactly what I needed.

Actually you can just use offset `0x3c` (`.e_lfanew`) straight away for your new overlay.  This is very much Microsoft-approved.

The "old-style" overlay offset is calculated from the `.e_cp` and `.e_cblp` shortwords at offsets `0x02` and `0x04`, as I said.  My modified CauseWay extender will now try to load the new payload (via the `0x3c` offset), then if that fails, load the old payload (via `0x02` and `0x04`).  So you can point `0x3c` at something that CauseWay does not understand &mdash; e.g. COFF.

> Though the requirement of a new tool looks unexpected. Maybe you'll consider making it a default (or optional via a switch) format of ia16-gcc.

The `lfanew` tool is still kind of experimental, so I am not sure about making it part of the whole toolchain's front-end just yet.

Thank you!

**Comment by stsp** [comment]
So you mean, even without lfanew
I can now just change 0x3c to my
offset?
Yes, it is possible to do.
But the availability of an lfanew offset
at 0x3c gives me a possibility to also
drop the parsing of a 3P header, so I'll
better use that and write a coff size to 0x1c.

**Comment by stsp** [comment]
I.e. if some extension allows me to
save 3-5 lines of code, I'll use that
extension. :)
lfanew allows me to save quite a bit,
so its a good extension, even if I could
otherwise get away with just using 0x3c
alone.

**Comment by tkchia** [comment]
Hello @stsp,

OK... here is a diagram that will hopefully clarify what you can (now) do:
```
        --------------           ---
0x0000 | e_magic "MZ" |           ^
       |--------------|           |
0x0002 | e_cblp       | ..        |
       |--------------|  : --     |
0x0004 | e_cp         | .:   |    |
       |--------------|      |    |
         ...                 |    | MZ portion
       |--------------|      |    |
0x003c | e_lfanew     | -------   |
       |--------------|      | |  |
0x0040 | ...relocs... |      | |  |
       |--------------|      | |  |
         ...                 | |  |
       | end of MZ    |      | |  v
       |--------------|      | | ---
0x9d2c | "3P"         | <----  |  ^
(e.g.) |--------------|        |  |
         ...                   |  | 3P portion
       | end of 3P    |        |  v
       |--------------|        | ---
0xa120 | 0x14c (COFF) | <------   ^
(e.g.) |--------------|           | COFF (e.g.) portion
         ...                      v
                                 ---
```

**Comment by tkchia** [comment]
So the offset at `0x3c` does _not_ need to point to the `3P` (CauseWay executable) payload &mdash; it can point to some other thing.

But yes, you (most probably) need to run `lfanew` on your CauseWay program, otherwise there will be no offset field at `0x3c` at all.

Thank you!

**Comment by stsp** [comment]
> But yes, you (most probably) need to run lfanew on your CauseWay program, otherwise there will be no offset field at 0x3c at all.

Yes, so if I understand correctly,
you propose me to write to 0x3c
this offset on my own, w/o lfanew.
The problem is that if I can't also
write to 0x1c (the size of coff), then
I will only set 0x3c to the end of
the coff payload, or otherwise I will
not be able to find that end.
Which is possible but doesn't allow
me to get rid of a 3P parsing.
In order to easily find both the start
and the end of the coff payload, I'd
better use lfanew and write a coff
size to 0x1c.

**Comment by tkchia** [comment]
Hello @stsp,

What is your proposed file format arrangement?  Perhaps you can draw a diagram to show what you are getting at.

I am probably missing something &mdash; I do not know why your loader code needs to know the _size_ of the COFF portion.  Certainly DJ Delorie's own stub does not need it.

Thank you!

**Comment by tkchia** [comment]
Hello @stsp,

This is also why I do not want to throw in extra options to `gcc-ia16` before they are ready for prime time.  Because I am never quite sure what exactly you want.

Thank you!

**Comment by stsp** [comment]
> Certainly DJ Delorie's own stub does not need it.

Because I extended the stub. :)
I am trying to write a replacement
of djgpp. djgpp never accepted any
patches (probably only from me)
and never available in fedora or
ubuntu. So I'll just write my own,
same as I did with freedos when
it made me angry enough (and
now we have fdpp).

So as to your question: current format
that I use is to read 0x3c to find coff,
and add 0x1c size to it to find the next
overlay, if any.
I could get away with just having 0x3c,
in which case it would point to the
coff end, and the coff start can be
found the old "hard" way. This is
possible, but lfanew makes things
much better.
